### PR TITLE
Initial ExecuteSQL and dataset transaction support

### PR DESCRIPTION
**Pull request by apjoseph**
This partially solves #131. I added support and corresponding tests for ExecuteSQL + transactions. I didn't address adding fields to keep the changeset relatively small.

### CODE REVIEW

**Comment by tbonfort** [comment]
For the failing tests, you'll need to fix the cmake build in .github/workflows/build-gdal.sh to `-DGDAL_USE_SQLITE3=ON`

I'm thinking that it would be more user friendly to have `ExecuteSQL` return a new `ResultSet` type instead of a `Layer`, in order to have it carry its own Close() method:

```go
type ResultSet {
 Layer
 ds *Dataset
}
func (ds *Dataset) ExecuteSQL(sql string, opts ...ExecuteSQLOption) (ResultSet, error) {
 /*your implementation except for the return type*/
}
func (rs Resultset) Close(opts ...CloseResultSetOption) error {
  /* your implementation from ReleaseResultSet */
}
```

thoughts ?

**Comment by apjoseph** [comment]
@tbonfort I very much prefer your suggested approach -much more idiomatic. I'll add that in once I can get the test runners to pass.

File: .github/workflows/build-gdal.sh [hunk 1]
```diff
@@ -27,7 +27,7 @@ if [ -f CMakeLists.txt ]; then
 		-DCMAKE_BUILD_TYPE=Release \
 		-DBUILD_TESTING=OFF \
 		-DGDAL_USE_CURL=OFF \
```
Comment (tbonfort) [inline .github/workflows/build-gdal.sh:29]: might also need to add `-DOGR_ENABLE_DRIVER_SQLITE=ON`

Comment (apjoseph) [reply .github/workflows/build-gdal.sh:29]: @tbonfort if it keeps failing, I'll see if I can find a less finicky transactional driver to use for testing.

Comment (tbonfort) [reply .github/workflows/build-gdal.sh:29]: build seems ok now. I'm not very keen on adding a spatialite dependency to the test-suite, are you able to rework your tests around that?

Comment (apjoseph) [reply .github/workflows/build-gdal.sh:29]: @tbonfort yeah I think I should be able to get it to work without libspatialite-dev.

File: godal.go [hunk 2]
```diff
@@ -3419,6 +3419,28 @@ type CloseResultSetOption interface {
...
+	}
+
+	if errNo != 0 || errMsg != "" {
+		clearUnhandledError()
+		return fmt.Errorf("unhandled error code %d: %s", errNo, errMsg)
+	}
+
+	return nil
+}
+
```
Comment (tbonfort) [inline godal.go:3443]: I'm not sure what you are trying to do here? You should make sure that no error is left unhandled, not reset it like this

File: godal.go [hunk 3]
```diff
@@ -3365,6 +3365,222 @@ func (ds *Dataset) LayerByName(name string) *Layer {
...
+// ExecuteSQL executes an SQL statement against the data store.
+func (ds *Dataset) ExecuteSQL(sql string, opts ...ExecuteSQLOption) (ResultSet, error) {
```
Comment (tbonfort) [inline godal.go:3445]: You should be returning a *ResultSet (and nil in case of error)

File: godal.go [hunk 4]
```diff
@@ -3365,6 +3365,222 @@ func (ds *Dataset) LayerByName(name string) *Layer {
...
+	geom *Geometry
+}
+
+func (sf SpatialFilter) setExecuteSQLOpt(eso *executeSQLOpts) {
+	eso.spatialFilter = sf
+}
+
+func NewSpatialFilter(geom *Geometry) SpatialFilter {
+	return SpatialFilter{geom}
+}
```
Comment (tbonfort) [inline godal.go:3378]: The API will be better worded with
```suggestion
type SpatialFilterOption struct {
	geom *Geometry
}

func (sf SpatialFilterOption) setExecuteSQLOpt(eso *executeSQLOpts) {
	eso.spatialFilter = sf
}

func SpatialFilter(geom *Geometry) SpatialFilterOption {
	return SpatialFilterOption{geom}
}
```

**Comment by tbonfort** [comment]
in 3584dce88e5188e3214f0f712f5d12cf9914450a you can find the proposed fixes

File: godal.go (hunk 2 above)
Comment (apjoseph) [reply godal.go]: @tbonfort thanks for pointing me in the right direction ! I had completely missed that I needed the backing c++ functions and was confused as to why errors were not being returned.

**Comment by tbonfort** [comment]
getting close. https://github.com/mattn/go-sqlite3/issues/1003#issuecomment-1017048443 might help for the remaining failure

**Comment by apjoseph** [comment]
@tbonfort I just changed it to a "SELECT 1".  I was originally using the SELECT on `sqlite_schema` to debug an issue where I was unable to SELECT against the test layer I previously created with CreateLayer -and simply running a SELECT on `sqlite_schema` to seemed to allow GDAL to recognize that the table exists. 

However it seems preemptively running any SQL statement beforehand corrects the problem. I would think the test table should be immediately accessible from ExecuteSQL -so I think this is bug with GDAL itself.

**Comment by tbonfort** [comment]
Given the test was only failing on ubuntu 20.04 and not on 22.04, I was under the impression the failure was rather due to the fact that the `sqlite_schema` table name was only available since more recent sqlite versions

**Comment by apjoseph** [comment]
@tbonfort that's correct. What I was referring to is the reason I put that statement in the test in the first place. 

If you remove:

```golang
	rs, err := ds.ExecuteSQL("SELECT 1", SQLiteDialect(), el)
```

then line 2581

```golang
      rs, err := ds.ExecuteSQL(ins, el, SQLiteDialect())

```

fails with:

```bash
=== RUN   TestExecuteSQL
    godal_test.go:2596: 
                Error Trace:    /home/ajoseph/repos/godal/godal_test.go:2596
                Error:          Received unexpected error:
                                In ExecuteSQL(): sqlite3_prepare_v2(INSERT INTO test VALUES (1,'POLYGON ((-72.573946 44.254648, -72.573946 44.255163, -72.573076 44.255163, -72.573076 44.254648, -72.573946 44.254648))'), (2,'POLYGON ((-72.576558 44.25799, -72.576558 44.258213, -72.576064 44.258213, -72.576064 44.25799, -72.576558 44.25799))')):
                                  no such table: test
                Test:           TestExecuteSQL
--- FAIL: TestExecuteSQL (0.01s)

```

To debug that failure I was running a check on `sqlite_schema` to see if the `test` table was actually registered (it was). Running that SELECT also allowed the insert to complete successfully. I am thinking that's probably indicative a bug in the OGR SQLite driver.

**Comment by tbonfort** [comment]
@rouault I'm wondering if we're using the OGR API correctly. Here is a reproducer of what's being done, translated to C code:

```c
#include <gdal.h>

int main(int argc, char **argv ) {
    GDALAllRegister();
    GDALDriverH drv = GDALGetDriverByName("SQLite");
    GDALDatasetH ds = GDALCreate(drv,"/vsimem/db.sqlite",0,0,0,GDT_Unknown, NULL);

    OGRLayerH lyr = GDALDatasetCreateLayer(ds,"test",NULL, wkbPolygon, NULL);

    GDALDatasetStartTransaction(ds, FALSE);
    GDALDatasetExecuteSQL(ds,"INSERT INTO test VALUES (1,'POLYGON ((-72.573946 44.254648, -72.573946 44.255163, -72.573076 44.255163, -72.573076 44.254648, -72.573946 44.254648))')",NULL, NULL);
    GDALDatasetRollbackTransaction(ds);
    printf("rollback\n");
    GDALDatasetStartTransaction(ds, FALSE);
    GDALDatasetExecuteSQL(ds,"INSERT INTO test VALUES (1,'POLYGON ((-72.573946 44.254648, -72.573946 44.255163, -72.573076 44.255163, -72.573076 44.254648, -72.573946 44.254648))')",NULL, NULL);
    GDALDatasetCommitTransaction(ds);
}
```
Is there a known reason why this code fails with:
```
rollback
ERROR 1: In ExecuteSQL(): sqlite3_prepare_v2(INSERT INTO test VALUES (1,'POLYGON ((-72.573946 44.254648, -72.573946 44.255163, -72.573076 44.255163, -72.573076 44.254648, -72.573946 44.254648))')):
  no such table: test
```
?. Thanks

**Comment by apjoseph** [comment]
@tbonfort  @rouault  Some observations:

- The insert will succeed without the `SELECT 1` if I also remove the `StartTransaction`
- The insert will fail  if I move the `SELECT 1`  when I am within the transaction.
- EmulatedTX has no impact 

This suggests to me that the SQLite driver is implicitly starting a transaction when executing `CreateLayer` and never actually closing its cursor. Source shows that [ExecuteSQL in SQLite does issue an explicit COMMIT](https://github.com/OSGeo/gdal/blob/835bd6dd0cea9970b1f121043bbc5740230dc362/ogr/ogrsf_frmts/sqlite/ogrsqliteexecutesql.cpp#L784) and since transactions do not nest without an explicit SAVEPOINT command, the previous statements would be committed. 

According to the [SQLite Docs](https://www.sqlite.org/lang_transaction.html):

> SQLite supports multiple simultaneous read transactions coming from separate database connections, possibly in separate threads or processes, but only one simultaneous write transaction.

> A read transaction is used for reading only. A write transaction allows both reading and writing. A read transaction is started by a SELECT statement, and a write transaction is started by statements like CREATE, DELETE, DROP, INSERT, or UPDATE (collectively "write statements"). **If a write statement occurs while a read transaction is active, then the read transaction is upgraded to a write transaction if possible. If some other database connection has already modified the database or is already in the process of modifying the database, then upgrading to a write transaction is not possible and the write statement will fail with [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy).**

> While a read transaction is active, any changes to the database that are implemented by separate database connections will not be seen by the database connection that started the read transaction. If database connection X is holding a read transaction, it is possible that some other database connection Y might change the content of the database while X's transaction is still open, however X will not be able to see those changes until after the transaction ends. While its read transaction is active, X will continue to see an historic snapshot of the database prior to the changes implemented by Y.

StartTransaction issues an explicit BEGIN so it would be seeing a snapshot of the database before the table is created. 


On Implicit transaction, the docs state:

> An implicit transaction (a transaction that is started automatically, not a transaction started by BEGIN) is committed automatically when the last active statement finishes. A statement **finishes when its last cursor closes**, which is guaranteed to happen when the prepared statement is [reset](https://www.sqlite.org/c3ref/reset.html) or [finalized](https://www.sqlite.org/c3ref/finalize.html). Some statements might "finish" for the purpose of transaction control prior to being reset or finalized, but there is no guarantee of this. The only way to ensure that a statement has "finished" is to invoke [sqlite3_reset()](https://www.sqlite.org/c3ref/reset.html) or [sqlite3_finalize()](https://www.sqlite.org/c3ref/finalize.html) on that statement. An open [sqlite3_blob](https://www.sqlite.org/c3ref/blob.html) used for incremental BLOB I/O also counts as an unfinished statement. The [sqlite3_blob](https://www.sqlite.org/c3ref/blob.html) finishes when it is [closed](https://www.sqlite.org/c3ref/blob_close.html).

>**The explicit COMMIT command runs immediately**, even if there are pending [SELECT](https://www.sqlite.org/lang_select.html) statements. However, if there are pending write operations, the COMMIT command will fail with an error code [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy).

So it seems the behavior we are seeing could only occur if a cursor created for CreateLayer remains open.

That same section suggests that we'd also see an SQLITE_BUSY error :

> An attempt to execute COMMIT might also result in an [SQLITE_BUSY](https://www.sqlite.org/rescode.html#busy) return code if an another thread or process has an open read connection. When COMMIT fails in this way, the transaction remains active and the COMMIT can be retried later after the reader has had a chance to clear.

I do not see any indication of this in the CPL_DEBUG output; but he sentence specifies "*might also result*" -which implies there can be situations in which this behavior does not occur.

**Comment by rouault** [comment]
> I'm wondering if we're using the OGR API correctly

@tbonfort I've a fix in https://github.com/OSGeo/gdal/pull/10509 *for the SQLite driver*
The issue is that a number of drivers implement a deferred table creation, because they need to create both the table and the fields in the same operation. ExecuteSQL() triggered this deferred creation, but this was a bit too late in the case of transactions, since the Rollback() would revert the table creation... I've also triggered the defered creation at StartTransaction() time. Other drivers might be impacted (the GPKG driver is from my quick testing, but the fix might be more involved since GPKG has also complex strategies for spatial index creation).

**Comment by rouault** [comment]
A workaround would be to issue a dummy ReleaseResultSet(ExecuteSQL("SELECT 1")) before StartTransaction

**Comment by apjoseph** [comment]
@rouault that's what I did in my test case to get it to pass
https://github.com/airbusgeo/godal/pull/132#issuecomment-2250404253 

Would it be helpful for me to create an issue for the SQLite driver problem in the GDAL repo or is it already a known bug?

**Pull request merged**